lib/Sema/NameBinding.cpp

```diff
@@ -36,60 +37,539 @@
 using namespace swift;
 
 //===----------------------------------------------------------------------===//
-// NameBinder
+// MARK: NameBinder and supporting types
 //===----------------------------------------------------------------------===//
 
 using ImportedModule = ModuleDecl::ImportedModule;
+using ImportedModuleDesc = SourceFile::ImportedModuleDesc;
 using ImportOptions = SourceFile::ImportOptions;
+using ImportFlags = SourceFile::ImportFlags;
+
+namespace {
+  /// Represents an import which the NameBinder knows exists, but which has not
+  /// yet had its options checked, module loaded, or cross-imports found.
+  ///
+  /// An UnboundImport may represent a physical ImportDecl written in the
+  /// source, or it may represent a cross-import overlay that has been found and
+  /// needs to be loaded.
+  struct UnboundImport {
+    /// If this UnboundImport directly represents an ImportDecl, the ImportDecl
+    /// it represents.
+    ///
+    /// This property should only be used to control where information is added
+    /// to the AST. Don't retrieve information directly from the ImportDecl; use
+    /// the other member variables.
+    NullablePtr<ImportDecl> ID;
+
+    SourceLoc importLoc;
+    ImportOptions options;
+    StringRef privateImportFileName;
+    Located<ImportKind> importKind;
+    ModuleDecl::AccessPathTy modulePath;
+    ModuleDecl::AccessPathTy declPath;
+
+    NullablePtr<DeclAttributes> attrs;
+
+    /// Create an UnboundImport for a user-written import declaration.
+    explicit UnboundImport(ImportDecl *ID);
+
+    /// Make sure the import is not a self-import.
+    bool checkNotTautological(const SourceFile &SF);
+
+    /// Make sure the module actually loaded, and diagnose if it didn't.
+    bool checkModuleLoaded(ModuleDecl *M, SourceFile &SF);
+
+    /// Find the top-level module for this module. If \p M is not a submodule,
+    /// returns \p M. If it is a submodule, returns either the parent module of
+    /// \p M or \c nullptr if the current module is the parent (which can happen
+    /// in a mixed-language framework).
+    NullablePtr<ModuleDecl> getTopLevelModule(ModuleDecl *M, SourceFile &SF);
```

Comment:
In the doc comment: "either the parent module" isn't quite right, I think, because the submodule may be nested and we will return the root of the hierarchy, not the immediate parent.

Reply:
Right—I was using "parent module" too loosely. New doc comment:
```
    /// Find the top-level module for this module; that is, if \p M is the
    /// module \c Foo.Bar.Baz, this finds \c Foo.
    ///
    /// Specifically, this method returns:
    ///
    /// \li \p M if \p M is a top-level module.
    /// \li \c nullptr if \p M is a submodule of \c SF's parent module. (This
    ///     corner case can occur in mixed-source frameworks, where Swift code
    ///     can import a Clang submodule of itself.)
    /// \li The top-level parent (i.e. ancestor with no parent) module above
    ///     \p M otherwise.
```